copy/single.go

```diff
@@ -29,26 +29,41 @@ import (
 
 // imageCopier tracks state specific to a single image (possibly an item of a manifest list)
 type imageCopier struct {
-	c                          *copier
-	manifestUpdates            *types.ManifestUpdateOptions
-	src                        *image.SourcedImage
-	diffIDsAreNeeded           bool
-	cannotModifyManifestReason string // The reason the manifest cannot be modified, or an empty string if it can
-	canSubstituteBlobs         bool
-	compressionFormat          *compressiontypes.Algorithm // Compression algorithm to use, if the user explicitly requested one, or nil.
-	compressionLevel           *int
+	c                             *copier
+	manifestUpdates               *types.ManifestUpdateOptions
+	src                           *image.SourcedImage
+	diffIDsAreNeeded              bool
+	cannotModifyManifestReason    string // The reason the manifest cannot be modified, or an empty string if it can
+	canSubstituteBlobs            bool
+	compressionFormat             *compressiontypes.Algorithm // Compression algorithm to use, if the user explicitly requested one, or nil.
+	compressionLevel              *int
+	requireCompressionFormatMatch bool
 }
 
 // copySingleImageResult carries data produced by copySingleImage
 type copySingleImageResult struct {
 	manifest         []byte
 	manifestMIMEType string
 	manifestDigest   digest.Digest
+	compressionAlgorithms  []compressiontypes.Algorithm
 }
 
+func removeDuplicateCompression(algos []compressiontypes.Algorithm) []compressiontypes.Algorithm {
```

Comment:
I think this can be built on `internal/set.Set` instead of manually.

Reply:
… and probably inlined in the two loops that collect the data.

Reply:
`Set.Add(` ensures that no duplicates are there ( I have added a test to verify that ) however `compression.Algorithm` is not a `comparable` hence I have created a `Set` using `compression.Algorithm.Name()` and its working fine.

Later I'm using `AlgorithmsByNames` to retrieve a slice of  `Algorithms`.